File: src/main/scala/search/engine/SearchEngine.scala
package search.engine

import java.nio.{ByteBuffer, ByteOrder}

object SearchEngine {

  sealed trait MultiSearchResult
  case class Found(foundAt: Int, foundNeedleId: Int) extends MultiSearchResult
  case object NotFound extends MultiSearchResult

  def indexOf(haystack: ByteBuffer, processor: UnrolledSearchProcessor): Int = {

    require(haystack.order == ByteOrder.LITTLE_ENDIAN)

    while (haystack.remaining >= 8) {
      val pos = processor.processUnrolled(haystack.getLong)
      if (pos >= 0) return haystack.position() - pos// - processor.needleLength
    }

    while (haystack.hasRemaining) {
      if (!processor.process(haystack.get)) return haystack.position() - processor.needleLength
    }

    -1
  }

  def indexOf(haystack: Array[Byte], processor: SearchProcessor, from: Int = 0): Int = {

    val matchAt = indexOfImpl(haystack, processor, from)

    if (matchAt < 0) -1
    else matchAt - processor.needleLength + 1
  }

  def indexOfMultiple(haystack: Array[Byte], processor: MultiSearchProcessor, from: Int = 0): MultiSearchResult = {

    val matchAt = indexOfImpl(haystack, processor, from)

    if (matchAt < 0) NotFound
    else Found(
      foundAt = matchAt - processor.needleLength + 1,
      foundNeedleId = processor.getFoundNeedleId)
  }

  private def indexOfImpl(haystack: Array[Byte], processor: SearchProcessor, from: Int): Int = {

    var i = from max 0

    while (i < haystack.length) {
      if (!processor.process(haystack(i))) return i
      i += 1
    }

    -1
  }

}

File: src/main/scala/search/engine/SearchEngine.java
package search.engine;

import search.algorithm.MultiSearchProcessor;
import search.algorithm.SearchProcessor;

public final class SearchEngine {

  private SearchEngine() { }

  public static int indexOf(byte[] haystack, SearchProcessor processor, int from) {

    final int matchAt = indexOfImpl(haystack, processor, from);

    if (matchAt < 0) return -1;
    else return matchAt - processor.needleLength() + 1;
  }

  public static MultiSearchResult indexOfMultiple(byte[] haystack, MultiSearchProcessor processor, int from) {

    final int matchAt = indexOfImpl(haystack, processor, from);

    return matchAt < 0 ? MultiSearchResult.NOT_FOUND : new MultiSearchResult(
        matchAt - processor.needleLength() + 1,
        processor.getFoundNeedleId());
  }

  private static int indexOfImpl(byte[] haystack, SearchProcessor processor, int from) {

    int i = Math.max(from, 0);

    while (i < haystack.length) {
      if (!processor.process(haystack[i])) return i;
      i++;
    }

    return -1;
  }

}
